fe/fe-core/src/main/java/org/apache/doris/datasource/InternalCatalog.java

```diff
@@ -1341,6 +1341,50 @@ public void replayCreateTable(String dbName, Table table) throws MetaNotFoundExc
         }
     }
 
+    public void addPartitionLike(Database db, String tableName, AddPartitionLikeClause addPartitionLikeClause)
+            throws DdlException {
+        try {
+            Table table = db.getTableOrDdlException(tableName);
+
+            if (table.getType() == TableType.VIEW) {
+                throw new DdlException("Not support create partition from a View");
+            }
+
+            table.readLock();
+            try {
+                if (table.getType().equals(TableType.OLAP)) {
+                    String partitionName = addPartitionLikeClause.getPartitionName();
+                    String existedName = addPartitionLikeClause.getExistedPartitionName();
+                    OlapTable olapTable = (OlapTable) table;
+                    Partition part = olapTable.getPartition(existedName);
+                    PartitionInfo partitionInfo = olapTable.getPartitionInfo();
+                    PartitionDesc partitionDesc = partitionInfo.toPartitionDesc((OlapTable) table);
+                    SinglePartitionDesc oldPartitionDesc = partitionDesc.getSinglePartitionDescByName(existedName);
+                    if (oldPartitionDesc == null) {
+                        throw new DdlException("Failed to ADD PARTITION" + partitionName + " LIKE "
+                                + existedName + ". Reason: " + "partition " + existedName + "not exist");
+                    }
+                    DistributionDesc distributionDesc = part.getDistributionInfo().toDistributionDesc();
+                    SinglePartitionDesc newPartitionDesc = new SinglePartitionDesc(false, partitionName,
+                            oldPartitionDesc.getPartitionKeyDesc(), oldPartitionDesc.getProperties());
+                    Map<String, String> properties = newPartitionDesc.getProperties();
+                    AddPartitionClause clause = new AddPartitionClause(newPartitionDesc, distributionDesc,
+                            properties, addPartitionLikeClause.getIsTempPartition());
+                    table.readUnlock();
```

Comment:
This `table.readUnlock();` should be in a `finally` block.
You split the whole logic into 2 parts:
One is for creating an `AddPartitionClause`, the other is for calling `addPartition()`.
And refactor the current `try..catch`